sigpy/mri/rf/slr.py

```diff
@@ -0,0 +1,648 @@
+        mp = (m > 0).astype(float)
+        mn = (m < 0).astype(float)
+
+        # design the positive and negative filters
+        c = np.exp(1j * 2 * np.pi / (2 * (N + 1))
+                   * np.concatenate([np.arange(0, N / 2 + 1, 1),
+                                     np.arange(-N / 2, 0, 1)]))
+        bp = signal.firls(N + 1, f, mp, w)  # the positive filter
+        bn = signal.firls(N + 1, f, mn, w)  # the negative filter
+
+        # combine the filters and demodulate
+        b = np.fft.ifft(np.multiply(np.fft.fft(bp - bn), c))
+        b = np.real(b[:N])
+        # hilbert transform to suppress negative passband
+        b = signal.hilbert(b)
+        # demodulate to DC
+        cShift = np.exp(-1j * 2 * np.pi / N * shift * np.arange(0, N, 1)) / 2 \
+            * np.exp(-1j * np.pi / N * shift)
+        b = np.multiply(b, cShift)
+
+    return b
+
+
+def dzgSliderrf(N=256, G=5, flip=np.pi / 2, phi=np.pi, tb=12,
+                d1=0.01, d2=0.01, cancelAlphaPhs=True):
+    bsf = np.sin(flip / 2)  # beta scaling factor
+
+    rf = np.zeros((N, G), dtype='complex')
+    for Gind in range(1, G + 1):
+        b = bsf * dzgSliderB(N, G, Gind, tb, d1, d2, phi)
+        rf[:, Gind - 1] = b2rf(b, cancelAlphaPhs)
+
+    return rf
+
+
+def b2rf(b, cancelAlphaPhs=False):
+    a = b2a(b)
+    if cancelAlphaPhs:
+        b = np.fft.ifft(np.fft.fft(b) *
+                        np.exp(-1j * np.angle(np.fft.fft(a[np.size(a)::-1]))))
+    rf = ab2rf(a, b)
+
+    return rf
+
+
+def b2a(b):
+    N = np.size(b)
+
+    Npad = N * 16
+    bcp = np.zeros(Npad, dtype=complex)
+    bcp[0:N:1] = b
+    bf = np.fft.fft(bcp)
+    bfmax = np.max(np.abs(bf))
+    if bfmax >= 1:
+        bf = bf / (1e-7 + bfmax)
+    afa = mag2mp(np.sqrt(1 - np.abs(bf) ** 2))
+    a = np.fft.fft(afa) / Npad
+    a = a[0:N:1]
+    a = a[::-1]
+
+    return a
+
+
+def mag2mp(x):
+    N = np.size(x)
+    xl = np.log(np.abs(x))  # Log of mag spectrum
+    xlf = np.fft.fft(xl)
+    xlfp = xlf
+    xlfp[0] = xlf[0]  # Keep DC the same
+    xlfp[1:(N // 2):1] = 2 * xlf[1:(N // 2):1]  # Double positive frequencies
+    xlfp[N // 2] = xlf[N // 2]  # keep half Nyquist the same
+    xlfp[N // 2 + 1:N:1] = 0  # zero negative frequencies
+    xlaf = np.fft.ifft(xlfp)
+    a = np.exp(xlaf)  # complex exponentiation
+
+    return a
+
+
+def ab2rf(a, b):
+    N = np.size(a)
+    rf = np.zeros(N, dtype=complex)
+
+    a = a.astype(complex)
+    b = b.astype(complex)
+
+    for ii in range(N - 1, -1, -1):
+
+        Cj = np.sqrt(1 / (1 + np.abs(b[ii] / a[ii]) ** 2))
+        Sj = np.conj(Cj * b[ii] / a[ii])
+        theta = np.arctan2(np.abs(Sj), Cj)
+        psi = np.angle(Sj)
+        rf[ii] = 2 * theta * np.exp(1j * psi)
+
+        # remove this rotation from polynomials
+        if ii > 0:
+            at = Cj * a + Sj * b
+            bt = -np.conj(Sj) * a + Cj * b
+            a = at[1:ii + 1:1]
+            b = bt[0:ii:1]
+
+    return rf
+
+
+def rootFlip(b, d1, flip, tb):
```

Comment:
`rootFlip` to `root_flip`. And it'll be great to put in a paper reference and documentation.